File: src/commands/budget.ts
import {
  BudgetConfig,
  NewBudgetConfig,
  BudgetSummary,
  BudgetAllocationWithCategory,
  BudgetAllocation,
  BudgetVsActual,
} from "@/lib/types";
import { getRunEnv, RUN_ENV, invokeTauri, invokeWeb } from "@/adapters";
import { logger } from "@/adapters";

export const getBudgetConfig = async (): Promise<BudgetConfig | null> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_budget_config");
      case RUN_ENV.WEB:
        return invokeWeb("get_budget_config");
      default:
        throw new Error("Unsupported");
    }
  } catch (error) {
    logger.error("Error fetching budget config.");
    throw error;
  }
};

export const upsertBudgetConfig = async (config: NewBudgetConfig): Promise<BudgetConfig> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("upsert_budget_config", { config });
      case RUN_ENV.WEB:
        return invokeWeb("upsert_budget_config", { config });
      default:
        throw new Error("Unsupported");
    }
  } catch (error) {
    logger.error("Error upserting budget config.");
    throw error;
  }
};

export const getBudgetSummary = async (): Promise<BudgetSummary> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_budget_summary");
      case RUN_ENV.WEB:
        return invokeWeb("get_budget_summary");
      default:
        throw new Error("Unsupported");
    }
  } catch (error) {
    logger.error("Error fetching budget summary.");
    throw error;
  }
};

export const getBudgetAllocations = async (): Promise<BudgetAllocationWithCategory[]> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_budget_allocations");
      case RUN_ENV.WEB:
        return invokeWeb("get_budget_allocations");
      default:
        throw new Error("Unsupported");
    }
  } catch (error) {
    logger.error("Error fetching budget allocations.");
    throw error;
  }
};

export const setBudgetAllocation = async (
  categoryId: string,
  amount: number,
): Promise<BudgetAllocation> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("set_budget_allocation", { categoryId, amount });
      case RUN_ENV.WEB:
        return invokeWeb("set_budget_allocation", { categoryId, amount });
      default:
        throw new Error("Unsupported");
    }
  } catch (error) {
    logger.error("Error setting budget allocation.");
    throw error;
  }
};

export const deleteBudgetAllocation = async (categoryId: string): Promise<number> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("delete_budget_allocation", { categoryId });
      case RUN_ENV.WEB:
        return invokeWeb("delete_budget_allocation", { categoryId });
      default:
        throw new Error("Unsupported");
    }
  } catch (error) {
    logger.error("Error deleting budget allocation.");
    throw error;
  }
};

export const getBudgetVsActual = async (month: string): Promise<BudgetVsActual> => {
  try {
    switch (getRunEnv()) {
      case RUN_ENV.DESKTOP:
        return invokeTauri("get_budget_vs_actual", { month });
      case RUN_ENV.WEB:
        return invokeWeb("get_budget_vs_actual", { month });
      default:
        throw new Error("Unsupported");
    }
  } catch (error) {
    logger.error("Error fetching budget vs actual.");
    throw error;
  }
};

File: src-tauri/src/commands/budget.rs
use std::sync::Arc;

use crate::{
    context::ServiceContext,
    events::{emit_resource_changed, ResourceEventPayload},
};
use log::debug;
use serde_json::json;
use tauri::{AppHandle, State};
use wealthfolio_core::budget::{
    BudgetAllocation, BudgetAllocationWithCategory, BudgetConfig, BudgetSummary, BudgetVsActual,
    NewBudgetConfig,
};

#[tauri::command]
pub async fn get_budget_config(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Option<BudgetConfig>, String> {
    debug!("Fetching budget config...");
    state
        .budget_service()
        .get_budget_config()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn upsert_budget_config(
    config: NewBudgetConfig,
    state: State<'_, Arc<ServiceContext>>,
    handle: AppHandle,
) -> Result<BudgetConfig, String> {
    debug!("Upserting budget config...");
    let result = state
        .budget_service()
        .upsert_budget_config(config)
        .await
        .map_err(|e| e.to_string())?;

    emit_resource_changed(
        &handle,
        ResourceEventPayload::new("budget", "updated", json!({ "config_id": result.id })),
    );

    Ok(result)
}

#[tauri::command]
pub async fn get_budget_summary(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<BudgetSummary, String> {
    debug!("Fetching budget summary...");
    state
        .budget_service()
        .get_budget_summary()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_budget_allocations(
    state: State<'_, Arc<ServiceContext>>,
) -> Result<Vec<BudgetAllocationWithCategory>, String> {
    debug!("Fetching budget allocations...");
    state
        .budget_service()
        .get_allocations()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn set_budget_allocation(
    category_id: String,
    amount: f64,
    state: State<'_, Arc<ServiceContext>>,
    handle: AppHandle,
) -> Result<BudgetAllocation, String> {
    debug!("Setting budget allocation...");
    let result = state
        .budget_service()
        .set_allocation(category_id.clone(), amount)
        .await
        .map_err(|e| e.to_string())?;

    emit_resource_changed(
        &handle,
        ResourceEventPayload::new(
            "budget_allocation",
            "updated",
            json!({ "category_id": category_id }),
        ),
    );

    Ok(result)
}

#[tauri::command]
pub async fn delete_budget_allocation(
    category_id: String,
    state: State<'_, Arc<ServiceContext>>,
    handle: AppHandle,
) -> Result<usize, String> {
    debug!("Deleting budget allocation...");
    let result = state
        .budget_service()
        .delete_allocation(&category_id)
        .await
        .map_err(|e| e.to_string())?;

    emit_resource_changed(
        &handle,
        ResourceEventPayload::new(
            "budget_allocation",
            "deleted",
            json!({ "category_id": category_id }),
        ),
    );

    Ok(result)
}

#[tauri::command]
pub async fn get_budget_vs_actual(
    month: String,
    state: State<'_, Arc<ServiceContext>>,
) -> Result<BudgetVsActual, String> {
    debug!("Fetching budget vs actual for month: {}", month);
    state
        .budget_service()
        .get_budget_vs_actual(&month)
        .map_err(|e| e.to_string())
}
